crates/matrix-sdk-ui/src/timeline/pinned_events_loader.rs

```diff
@@ -0,0 +1,221 @@
+                    let semaphore = Arc::clone(&semaphore);
+                    let provider = Arc::clone(&provider);
+                    async move {
+                        let permit = semaphore
+                            .acquire()
+                            .await
+                            .map_err(|_| PinnedEventsLoaderError::SemaphoreNotAcquired)?;
+                        let ret = provider
+                            .room_event(&id)
+                            .await
+                            .map_err(|_| PinnedEventsLoaderError::EventNotFound(id.to_owned()));
+                        drop(permit);
+                        ret
+                    }
+                }));
+            }
+
+            for handle in handles {
+                if let Ok(Ok(ev)) = handle.await {
+                    loaded_events.push(ev)
+                }
+            }
+        }
+
+        self.cache.set_bulk(&loaded_events).await;
+
+        fn timestamp(item: &SyncTimelineEvent) -> MilliSecondsSinceUnixEpoch {
+            item.event
+                .deserialize()
+                .map(|e| e.origin_server_ts())
+                .unwrap_or_else(|_| MilliSecondsSinceUnixEpoch::now())
+        }
+
+        let sorted_events = loaded_events
+            .into_iter()
+            .sorted_by(|e1, e2| timestamp(e1).cmp(&timestamp(e2)))
+            .collect();
+
+        Ok(sorted_events)
+    }
+
+    /// Updates the cache with the provided events if they're associated with a
+    /// pinned event id, then reloads the list of pinned events if there
+    /// were any changes.
+    ///
+    /// Returns an `Option` with either a new list of events if any of them
+    /// changed in this update or `None` if they didn't.
+    pub async fn update_if_needed(
+        &self,
+        events: Vec<impl Into<SyncTimelineEvent>>,
+    ) -> Option<Vec<SyncTimelineEvent>> {
+        let mut to_update = Vec::new();
+        for ev in events {
+            let ev = ev.into();
+            if let Some(ev_id) = ev.event_id() {
+                if self.room.room_is_pinned_event_id(&ev_id) {
+                    to_update.push(ev);
+                }
+            }
+        }
+
+        if !to_update.is_empty() {
+            self.cache.set_bulk(&to_update).await;
+            self.load_events().await.ok()
+        } else {
+            None
+        }
+    }
+}
+
+#[cfg_attr(target_arch = "wasm32", async_trait::async_trait(?Send))]
+#[cfg_attr(not(target_arch = "wasm32"), async_trait::async_trait)]
+pub trait PinnedEventsRoom: SendOutsideWasm + SyncOutsideWasm {
+    /// Load a single room event.
+    async fn room_event(&self, event_id: &EventId) -> Result<SyncTimelineEvent, PaginatorError>;
+
+    /// Get the pinned event ids for a room.
+    fn room_pinned_event_ids(&self) -> Vec<OwnedEventId>;
+
+    /// Checks whether an event id is pinned in this room.
+    fn room_is_pinned_event_id(&self, event_id: &EventId) -> bool;
```

Comment:
I think just `is_pinned_event_id` or `is_pinned_event`, or even `is_pinned`.

Reply:
(Basically I think `room` is not helpful here.

Reply:
The problem with these is there already are some existing methods in `Room` that have these names and they access private properties, and then I had to create new methods to expose them in the `PinnedEventsRoom` trait and use them in this file. 

I can't just replace the existing methods because they need access to those private properties to work and they can't have the same name because the compiler will mix them up. Unless there's some way to distinguish between those?

A good example of this is `PinnedEventsRoom::room_pinned_event_ids` and `Room::pinned_event_ids`.

Reply:
Ok, I think I got it with https://github.com/matrix-org/matrix-rust-sdk/pull/3773/commits/237ee9ae0e3bef053cedd621b8a83fd2e9a29e7d.

Reply:
Looks good.